src/main/kotlin/org/rust/ide/annotator/CargoCheckExtAnnotator.kt

```diff
@@ -0,0 +1,630 @@
+import com.intellij.openapi.util.TextRange
+import com.intellij.openapi.vfs.VirtualFileManager
+import com.intellij.openapi.vfs.newvfs.BulkFileListener
+import com.intellij.openapi.vfs.newvfs.events.VFileEvent
+import com.intellij.openapi.diagnostic.Logger as IJLogger
+
+import com.google.gson.*
+
+import org.apache.commons.lang.StringEscapeUtils.escapeHtml
+
+object IJDebug {
+    val defaultTitle = "CARGO-CHECK"
+
+    fun log(msg : String, title : String = defaultTitle) {
+        IJLogger.getInstance(title).debug(msg)
+    }
+
+}
+
+object CargoJson {
+
+    private fun<T> checkExp(name : String, actual : T, expected : T, forContext : String = "") {
+        if (actual == expected)
+            return
+
+        val ctxStr = if (forContext != "") "for $forContext " else ""
+        val err = "Unexpected $name `$actual` $ctxStr(expected `$expected`)"
+
+        throw RuntimeException(err)
+    }
+
+    data class Target(
+        val kind: List<String>,
+        val name : String,
+        val src_path : String)
+
+    @Suppress("UNUSED")
+    //These are special compiler messages that we're going to ignore.
+    data class CompilerMessage(
+        val features : JsonObject,
+        val filenames : List<String>,
+        val package_id: String,
+        val profile : Profile,
+        val reason : String,
+        val target : Target
+        ) {
+        data class Profile(
+            val debug_assertions : Boolean,
+            val debuginfo : Int,
+            val opt_level : String,
+            val test : Boolean
+            )
+    }
+
+    //A top level compiler message
+    data class TopMessage(
+        val message: Message,
+        @Suppress("UNUSED")
+        val package_id : String,
+        val reason : String,
+        val target : Target
+    ) {
+        init {
+            checkExp("reason", reason, "compiler-message")
+            checkExp("target.kind", target.kind, listOf("bin"))
+        }
+
+        abstract class InfoMessageBase(override val message: String, override val spans : List<Span>) : Message() {
+
+            override fun doAnnotate(holder: AnnotationHolder, target: Target) {
+                //We don't expect these to be called directly.
+                //We only expect help messages as children of error messages
+
+            }
+        }
+
+        class HelpMessage(message: String, spans: List<Span>) : InfoMessageBase(message, spans)
+        class NoteMessage(message: String, spans: List<Span>) : InfoMessageBase(message, spans)
+
+        //Used for both error and warning
+        abstract class ErrorBaseMessage(
+            val children: List<Message>,
+            val code: Code?,
+            override val message: String,
+            override val spans: List<Span>,
+            val level : String //Only "warning" and "error" allowed
+        ) : Message() {
+            init {
+                children.forEach { assert(it is NoteMessage || it is HelpMessage) }
```

Comment:
It's preferable to use `check` instead of `assert`, because assertions are usually disabled.